bindings/apns/apns.go

```diff
@@ -0,0 +1,160 @@
+	expirationKey     = "apns-expiration"
+	keyIDKey          = "key-id"
+	messageIDKey      = "apns-id"
+	priorityKey       = "apns-priority"
+	privateKeyKey     = "private-key"
+	productionPrefix  = "https://api.push.apple.com/3/device/"
+	pushTypeKey       = "apns-push-type"
+	teamIDKey         = "team-id"
+	topicKey          = "apns-topic"
+)
+
+type notificationResponse struct {
+	MessageID string `json:"messageID"`
+}
+
+type errorResponse struct {
+	Reason    string `json:"reason"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+// APNS implements an outbound binding that allows services to send push
+// notifications to Apple devices using Apple's Push Notification Service.
+type APNS struct {
+	logger               logger.Logger
+	client               *http.Client
+	urlPrefix            string
+	authorizationBuilder *authorizationBuilder
+}
+
+// Creates a new output binding instance for the Apple Push Notification
+// Service.
+func NewAPNS(logger logger.Logger) *APNS {
+	return &APNS{
+		logger: logger,
+		client: &http.Client{},
+		authorizationBuilder: &authorizationBuilder{
+			logger: logger,
+			mutex:  sync.RWMutex{},
+		},
+	}
+}
+
+func (a *APNS) Init(metadata bindings.Metadata) error {
+	if err := a.makeURLPrefix(metadata); err != nil {
+		return err
+	}
+
+	if err := a.extractKeyID(metadata); err != nil {
+		return err
+	}
+
+	if err := a.extractTeamID(metadata); err != nil {
+		return err
+	}
+
+	return a.extractPrivateKey(metadata)
+}
+
+func (a *APNS) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{bindings.CreateOperation}
+}
+
+func (a *APNS) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	if req.Operation != bindings.CreateOperation {
+		return nil, fmt.Errorf("operation not supported: %v", req.Operation)
+	}
+
+	op := &createOperation{
+		authorizationBuilder: a.authorizationBuilder,
+		client:               a.client,
+		logger:               a.logger,
+		req:                  req,
+		urlPrefix:            a.urlPrefix,
+	}
+	return op.run()
+}
+
+func (a *APNS) makeURLPrefix(metadata bindings.Metadata) error {
+	if value, ok := metadata.Properties[developmentKey]; ok && value != "" {
+		switch value {
+		case "true":
+			a.logger.Debug("Using the development APNS service")
+			a.urlPrefix = developmentPrefix
+
+		case "false":
+			a.logger.Debug("Using the production APNS service")
+			a.urlPrefix = productionPrefix
+
+		default:
+			return fmt.Errorf(
+				"invalid value for development parameter: %v",
+				value,
+			)
+		}
+	} else {
+		a.logger.Debug("Using the production APNS service")
+		a.urlPrefix = productionPrefix
+	}
+
+	return nil
+}
+
+func (a *APNS) extractKeyID(metadata bindings.Metadata) error {
+	if value, ok := metadata.Properties[keyIDKey]; ok && value != "" {
+		a.authorizationBuilder.keyID = value
+	} else {
+		return errors.New("the key-id parameter is required")
+	}
+
+	return nil
+}
+
+func (a *APNS) extractTeamID(metadata bindings.Metadata) error {
```

Comment:
Would this be more readable? 

```go
func (a *APNS) extractKeyID(metadata bindings.Metadata) error {
	if value, ok := metadata.Properties[keyIDKey]; ok && value != "" {
		a.authorizationBuilder.keyID = value
                return nil
	}
	return errors.New("the key-id parameter is required")
}
```

Reply:
Yes. I agree. I corrected it in the code.